lib/charms/vault_k8s/v0/vault_kv.py

```diff
@@ -0,0 +1,370 @@
+        self.unit.status = BlockedStatus("Waiting for vault-kv relation")
+
+    def _on_ready(self, event: vault_kv.VaultKvReadyEvent):
+        relation = self.model.get_relation(event.relation_name, event.relation_id)
+        if relation is None:
+            return
+        vault_url = self.interface.vault_url(relation)
+        mount = self.interface.mount(relation)
+
+        unit_credentials = self.interface.unit_credentials(relation)
+        # unit_credentials is a juju secret id
+        secret = self.model.get_secret(id=unit_credentials)
+        secret_content = secret.get_content(refresh=True)
+        role_id = secret_content["role-id"]
+        role_secret_id = secret_content["role-secret-id"]
+
+        self._configure(vault_url, mount, role_id, role_secret_id)
+
+        self.unit.status = ActiveStatus()
+
+    def _on_gone_away(self, event: vault_kv.VaultKvGoneAwayEvent):
+        self.unit.status = BlockedStatus("Waiting for vault-kv relation")
+
+    def _configure(self, vault_url: str, mount: str, role_id: str, role_secret_id: str):
+        pass
+
+
+if __name__ == "__main__":
+    main(ExampleRequirerCharm)
+
+You can integrate both charms by running:
+
+```bash
+juju integrate <vault provider charm> <vault requirer charm>
+```
+"""
+
+import json
+import logging
+from typing import Any, Dict, Optional
+
+import ops
+
+logger = logging.getLogger(__name__)
+
+
+# The unique Charmhub library identifier, never change it
+LIBID = "to_fill"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+
+class NewVaultKvClientAttachedEvent(ops.EventBase):
+    """Ready VaultKvClients Event."""
+
+    def __init__(
+        self,
+        handle: ops.Handle,
+        relation_id: int,
+        relation_name: str,
+        mount_suffix: str,
+    ):
+        super().__init__(handle)
+        self.relation_id = relation_id
+        self.relation_name = relation_name
+        self.mount_suffix = mount_suffix
+
+    def snapshot(self) -> dict:
+        """Return snapshot data that should be persisted."""
+        return {
+            "relation_id": self.relation_id,
+            "relation_name": self.relation_name,
+            "mount_suffix": self.mount_suffix,
+        }
+
+    def restore(self, snapshot: Dict[str, Any]):
+        """Restore the value state from a given snapshot."""
+        super().restore(snapshot)
+        self.relation_id = snapshot["relation_id"]
+        self.relation_name = snapshot["relation_name"]
+        self.mount_suffix = snapshot["mount_suffix"]
+
+
+class VaultKvProviderEvents(ops.ObjectEvents):
+    """List of events that the Vault Kv provider charm can leverage."""
+
+    new_vault_kv_client_attached = ops.EventSource(NewVaultKvClientAttachedEvent)
+
+
+class VaultKvProvides(ops.Object):
```

Comment:
I don't believe other charms are expected to implement the Provides interface, should we move this class (and related event) from this lib ?

Reply:
Humm, we may want it in the machine charm for vault at some point, I'm fine if it stays here